prow/interrupts/interrupts.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package interrupts exposes helpers for graceful handling of interrupt signals
+package interrupts
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// NewManager returns a reference to the singleton manager
+func NewManager() Manager {
+	if single == nil {
+		m := sync.Mutex{}
+		single = &manager{
+			c:  sync.NewCond(&m),
+			wg: sync.WaitGroup{},
+		}
+		go single.handleInterrupt()
+	}
+
+	return single
+}
+
+// Manager knows how to run tasks with graceful termination on interrupts
+type Manager interface {
+	ListenAndServe(server *http.Server, gracePeriod time.Duration)
+	ListenAndServeTLS(server *http.Server, certFile, keyFile string, gracePeriod time.Duration)
+	Run(work func(ctx context.Context) error)
+	Tick(work func(), interval time.Duration)
+	Context() context.Context
+	WaitForGracefulShutdown()
+}
+
+type manager struct {
+	// only one signal handler should be installed, so we use a cond to
+	// broadcast to workers that an interrupt has occurred
+	c *sync.Cond
+	// we want to ensure that all registered servers and workers get a
+	// change to gracefully shut down
+	wg sync.WaitGroup
+}
+
+// handleInterrupt turns an interrupt into a broadcast for our condition.
+// This must be called _first_ before any work is registered with the
+// manager, or there will be a deadlock.
+func (m *manager) handleInterrupt() {
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, syscall.SIGTERM, syscall.SIGABRT)
+	m.c.L.Lock()
```

Comment:
This will prevent anything else from waiting on the condition variable because the lock needs to be held to call `Wait()`.
The lock doesn't need to be held here so I think this can just be removed.